src/services/db.ts

```diff
@@ -118,6 +224,99 @@ export class WefwefDB extends Dexie {
       .limit(limit)
       .toArray();
   }
+
+  /*
+   * Settings
+   */
+
+  async populateDefaultSettings(tx: Transaction) {
+    const settingsTable = tx.table("settings");
+    settingsTable.bulkAdd(defaultSettings);
+  }
+
+  async migrateFromLocalStorageSettings(tx: Transaction) {
+    const localStorageMigrationKeys = {
+      font_size_multiplier: "appearance--font-size-multiplier",
+      use_system_font_size: "appearance--font-use-system",
+      collapse_comment_threads: "appearance--collapse-comment-threads",
+      post_appearance_type: "appearance--post-type",
+      use_system_dark_mode: "appearance--dark-use-system",
+      user_dark_mode: "appearance--dark-user-mode",
+    };
+
+    const settingsTable = tx.table("settings");
+
+    return await Promise.all(
+      Object.entries(localStorageMigrationKeys).map(
+        ([key, localStorageKey]) => {
+          const localStorageValue = localStorage.getItem(localStorageKey);
+
+          if (!localStorageValue) {
+            return Promise.resolve();
+          }
+
+          return settingsTable
+            .where(CompoundKeys.settings.key_and_user_handle_and_community)
+            .equals([key, "", ""])
+            .modify({
+              value: JSON.parse(localStorageValue),
+            });
+        }
+      )
+    );
+  }
+
+  async getSetting<T extends keyof SettingValueTypes>(
+    key: T,
+    specificity?: {
+      user_handle?: string;
+      community?: string;
+    }
+  ) {
+    const { user_handle = "", community = "" } = specificity || {};
+
+    const query = this.settings
+      .where(CompoundKeys.settings.key_and_user_handle_and_community)
+      .equals([key, user_handle, community]);
```

Comment:
Will there be settings where there is a default with the possibility of selecting something different per user? Maybe defaulting to the query with community empty followed by user and community empty would be good.

Reply:
@rsammelson this query supports all possible combinations for a setting.
- without user and without community
- with user and community
- with user but without community
- without user but with community

So if you have 2 accounts and you change such setting in one, you'll still have the default in the other one. You can also change the default for every account in theory with this but not sure how the UX would be.

Reply:
I mean does the function automatically fall through, if so I just don't understand how the query works.

Reply:
``` typescript
const query = this.settings
      .where(CompoundKeys.settings.key_and_user_handle_and_community)
      .equals([key, user_handle, community]);
```
the compound key in this `where` is `"[key+user_handle+community]"`. So if you
- give a "key" but leave handle and community empty; parameter to `equals` would be: `["key", "", ""]` which would match the setting which has the given "key" but doesn't have a community or handle.
- give "key" and a "handle";  parameter to `equals` would be: `["key", "handle", ""]` which would match the setting with the given "key" and "handle" but doesn't have a community. 

There can be multiple rows with the same "key" but with different handle or community values.

Reply:
I guess I wasn't clear, I meant even if you pass in all three it might be convenient if it automatically checks for an entry that exists with only some set if there is none with all set. Does that make any sense?

Reply:
Oh I see.  Yeah it probably makes sense t o have some sort of fall back mechanism 👍

Reply:
@rsammelson added the fallback mechanism. It  tries in the following order:
- key, handle and community matches (user setting for a specific community)
- key, handle (if provided) matches but community is empty (user setting without a specific community)
- key, community (if provided) matches but handle is empty (community setting without a user)
- just the key matches (global setting)